adafruit_ble/hid.py

```diff
@@ -0,0 +1,269 @@
+ # The MIT License (MIT)
+#
+# Copyright (c) 2019 Dan Halbert for Adafruit Industries
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+"""
+`adafruit_ble.hid_keyboard`
+====================================================
+
+BLE HID
+
+* Author(s): Dan Halbert for Adafruit Industries
+
+"""
+import struct
+
+from bleio import Attribute, Characteristic, Descriptor, Peripheral, Service, UUID
+from .advertising import ServerAdvertisement
+
+class HID:
+    """
+    Provide devices for HID over BLE.
+
+    :param str name: Name to advertise for server. If None, use default Peripheral name.
+
+    Example::
+
+        from adafruit_ble.hid import HID
+
+        hid = HID()
+    """
+
+    HUMAN_INTERFACE_DEVICE_UUID = UUID(0x1812)
+    REPORT_UUID = UUID(0x2A4D)
+    REPORT_MAP_UUID = UUID(0x2A4B)
+    HID_INFORMATION_UUID = UUID(0x2A4A)
+    HID_CONTROL_POINT_UUID = UUID(0x2A4C)
+    REPORT_REF_DESCR_UUID = UUID(0x2908)
+    _REPORT_TYPE_INPUT = 1
+    # Boot keyboard and mouse not currently supported.
+    # PROTOCOL_MODE_UUID = UUID(0x2A4E)
+    # HID_BOOT_KEYBOARD_INPUT_REPORT_UUID = UUID(0x2A22)
+    # HID_BOOT_KEYBOARD_OUTPUT_REPORT_UUID = UUID(0x2A32)
+    # HID_BOOT_MOUSE_INPUT_REPORT_UUID = UUID(0x2A33)
+
+    #pylint: disable=line-too-long
+    HID_DESCRIPTOR = (
```

Comment:
Is this identical to a USB HID descriptor?

Reply:
Yes, it is. This descriptor is taken more or less intact from the Arduino Bluefruit library.